### PR TITLE
Integrate BNF Article Import/Export Functionality. DDFHER-164

### PR DESCRIPTION
**Pull request by rasben**
Please review commit by commit - I've tried my best to break it up, and describe it well here, and any summary I write here will just be confusing.

Locally, I've been testing with just enabling both `bnf_client` and `bnf_server`. As one of my `@todo`'s say, we need to figure out how to actually work with this locally.
I've hardcoded that the server is just the same as the current site. This obviously also need to change :)

I'm also unsure about the authentication. I've added a few `@todo`'s there with questions.
It has hardcoded username and password that only works locally, but, I didn't want to spend energy making it "proper" if we ended up doing something completely different.

In conclusion - this works, but, it's not necessarily what I had in mind when I started out. I'm a bit unsure about the GraphQL mutation, but still - it works as intended.

I'd love some help with my `@todo`'s, but, I actually think I'd prefer if we just talk in person about those - maybe some pair-programming? :)

Feel free to test locally by enabling `bnf_client` and `bnf_server`.

Screenshot 2024-12-10 at 21 32 24

### CODE REVIEW

File: web/modules/custom/bnf/bnf.info.yml [hunk 1]
```diff
@@ -2,6 +2,4 @@ name: 'BNF Common'
 type: module
 description: 'Common code between BNF (Bibliotekernes Nationale Formidling) client and server modules'
 package: 'DPL - BNF'
-# It doesn't make sense to enable this module on its own.
```
Comment (xendk) [inline web/modules/custom/bnf/bnf.info.yml]: Then put `$settings['extension_discovery_scan_tests'] = TRUE` in your settings file.

**Review by xendk — CHANGES_REQUESTED** [review]

File: config/sync/core.entity_form_display.node.article.default.yml [hunk 2]
```diff
@@ -3,6 +3,7 @@ langcode: en
 status: true
 dependencies:
   config:
+    - field.field.node.article.field_bnf_uuid
```
[the remark this answers is not in the document]
Comment (rasben) [reply config/sync/core.entity_form_display.node.article.default.yml:6]: But, this is the UUID field that we'll use across sites, to keep track of which content has been imported already

[the remark this answers is not in the document]
Comment (xendk) [reply config/sync/core.entity_form_display.node.article.default.yml:6]: Oh right, upstream uuid.

File: web/modules/custom/bnf/bnf_server/src/Plugin/GraphQL/DataProducer/ImportRequestProducer.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+   * @param string $callbackUrl
+   *   The callback URL.
+   *
+   * @return string[]
+   *   The response data.
+   */
+  public function resolve(string $uuid, string $callbackUrl): array {
+    // For now, we only support articles. In the future, this should be
+    // sent along as a parameter, as GraphQL exposes different queries
+    // for each node type (nodeArticle)
```
[the remark this answers is not in the document]
Comment (rasben) [reply web/modules/custom/bnf/bnf_server/src/Plugin/GraphQL/DataProducer/ImportRequestProducer.php:73]: I've asked in Zulip, we'll see what they say :)

File: web/modules/custom/bnf/bnf_server/src/Plugin/GraphQL/DataProducer/ImportRequestProducer.php [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+use Psr\Log\LoggerInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Resolves the `importRequest` mutation.
+ *
+ * @DataProducer(
+ *   id = "import_request_producer",
+ *   name = "Import Request Producer",
+ *   description = "Handles the importRequest mutation.",
```
[the remark this answers is not in the document]
Comment (rasben) [reply web/modules/custom/bnf/bnf_server/src/Plugin/GraphQL/DataProducer/ImportRequestProducer.php:21]: I've updated the documentation, hopefully you feel it's sufficient :)

File: web/modules/custom/bnf/bnf_server/src/Plugin/GraphQL/DataProducer/ImportRequestProducer.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+namespace Drupal\bnf_server\Plugin\GraphQL\DataProducer;
+
+use Drupal\bnf\Services\BnfImporter;
+use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
+use Drupal\graphql\Plugin\GraphQL\DataProducer\DataProducerPluginBase;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Resolves the `importRequest` mutation.
```
[the remark this answers is not in the document]
Comment (rasben) [reply web/modules/custom/bnf/bnf_server/src/Plugin/GraphQL/DataProducer/ImportRequestProducer.php]: I agree with you, but for me, it's also a bit magic - I'm not sure I 100% understand it, which my poor documentation reflects

I've tried to expand it now - hopefully you're satisfied :)

File: web/modules/custom/bnf/src/Services/BnfImporter.php [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    $existing_nodes =
+      $node_storage->loadByProperties([self::UUID_FIELD => $uuid]);
+
+    if (!empty($existing_nodes)) {
+      $this->logger->error(
+        'Cannot import @type @uuid from @url - Node already exists.',
+        ['@type' => $node_type, '@uuid' => $uuid, '@url' => $endpoint_url]
+      );
+
+      throw new \Exception((string) $this->translation->translate(
```
[the remark this answers is not in the document]
Comment (rasben) [reply web/modules/custom/bnf/src/Services/BnfImporter.php:46]: According to PHPStorm, AlreadyExistsException doesn't exist in Drupal - it's an external package.
Are you sure that's what you meant?

But yeah, this is used as a way to bubble information down, and to the external site, to display the message for the editors - hence why I want it to be translated.

File: web/modules/custom/bnf/bnf_client/bnf_client.module [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  // When adding the submit handler here, it happens after ::save() and any
+  // validation.
+  $form['actions']['submit']['#submit'][] = 'bnf_client_form_node_form_submit';
+
+  // Let's hide the publishing button when BNF is checked, as the value will
+  // be ignored and the node will be published regardless.
+  $form['status']['#states']['invisible'] =
+    [':input[name="bnf"]' => ['checked' => TRUE]];
+
+  $form['bnf'] = [
```
[the remark this answers is not in the document]
Comment (rasben) [reply web/modules/custom/bnf/bnf_client/bnf_client.module:42]: I agree, but I have a hard time seeing what we can do differently. Any suggestions? Do you consider this a blocker?

File: config/sync/core.entity_form_display.node.article.default.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (xendk) [reply config/sync/core.entity_form_display.node.article.default.yml:6]: But I don't like that it's just a regular field. It makes a hard coupling between the module and the site configuration. Never mind the fact that we don't need it on the server.

Sadly, I don't think it's possible to hook into entities base field definition. Have you looked into what contrib modules that need to add data to entities does?

File: web/modules/custom/bnf/bnf.services.yml [hunk 8]
```diff
@@ -0,0 +1,12 @@
+---
+services:
+  bnf.importer:
+    class: Drupal\bnf\Services\BnfImporter
+    arguments:
```
[the remark this answers is not in the document]
Comment (rasben) [reply web/modules/custom/bnf/bnf.services.yml:5]: We don't seem to use autowire in the rest of the project - should i still implement it?

File: web/modules/custom/bnf/bnf_server/src/Plugin/GraphQL/DataProducer/ImportRequestProducer.php [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+ *     )
+ *   }
+ * )
+ */
+class ImportRequestProducer extends DataProducerPluginBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
```
[the remark this answers is not in the document]
Comment (rasben) [reply web/modules/custom/bnf/bnf_server/src/Plugin/GraphQL/DataProducer/ImportRequestProducer.php:43]: We don't seem to use autowire in the rest of the project - should i still implement it?

File: config/sync/core.entity_form_display.node.article.default.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (rasben) [reply config/sync/core.entity_form_display.node.article.default.yml:6]: >Never mind the fact that we don't need it on the server.

But.. we do need it on the server? It will be used when the libraries pull content from BNF?

>Sadly, I don't think it's possible to hook into entities base field definition. Have you looked into what contrib modules that need to add data to entities does?

I actually feel better about doing it this way, than adding a new dependency on another contrib module - but, if you have a suggestion for a field, I'm willing to hear it :)

File: docker-compose.yml [hunk 10]
```diff
@@ -31,6 +31,10 @@ x-environment:
     LAGOON_ENVIRONMENT: 'local'
     LAGOON_ENVIRONMENT_TYPE: ${LAGOON_ENVIRONMENT_TYPE:-local}
     WEBROOT: web
+    GRAPHQL_USER_NAME: graphql_consumer
+    GRAPHQL_USER_PASSWORD: test
+    # @todo - what do we set this to?
+    BNF_SERVER_GRAPHQL_ENDPOINT: "https://localhost/graphql"
```
Comment (rasben) [inline docker-compose.yml:37]: I'm not sure what we do here - this doesnt work.
I've tested locally, and if i set it up to my own test domain (https://varnish.dpl-cms.orb.local/graphql) all is well

File: config/sync/core.entity_form_display.node.article.default.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (xendk) [reply config/sync/core.entity_form_display.node.article.default.yml:6]: > It will be used when the libraries pull content from BNF

Server side, it's just the node UUID. What we need is an identifier that identifies a node on the server. As nodes are born with a UUID, we already have that, hell, as it only needs to be unique on the server we could use `nid`.

But if we sync UUID when we sync nodes, we don't even have to store the "original UUID" as the same content on different sites will have the same UUID (`nid` wont work for this). The only data we then need to store is the fact that this node was imported/exported, and that's just to be able to make a list of imported content.

> I actually feel better about doing it this way, than adding a new dependency on another contrib module - but, if you have a suggestion for a field, I'm willing to hear it :)

Wasn't suggesting you find a contrib module that does it for you, but that you looked into *how* they do it. `scheduler` would be a good candidate, it needs to add dates to all nodes, but they're not fields in the configuration.

File: web/modules/custom/bnf/bnf_server/src/Plugin/GraphQL/DataProducer/ImportRequestProducer.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (xendk) [reply web/modules/custom/bnf/bnf_server/src/Plugin/GraphQL/DataProducer/ImportRequestProducer.php:21]: Good enough for government work.

I'd have left out the description of the parameters, they're described on the `resolve()` method anyway.

File: web/modules/custom/bnf/bnf_server/src/Plugin/GraphQL/DataProducer/ImportRequestProducer.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (xendk) [reply web/modules/custom/bnf/bnf_server/src/Plugin/GraphQL/DataProducer/ImportRequestProducer.php:43]: Of course. That technical debt exists in other parts of the project is no excuse for adding more. And if I was working on some existing code, I would add it in.

File: web/modules/custom/bnf/src/Services/BnfImporter.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (xendk) [reply web/modules/custom/bnf/src/Services/BnfImporter.php:47]: > Are you sure that's what you meant?

Yes. What I meant: Make your own. 

```php
<?php

declare(strict_types=1);

namespace Drupal\bnf\Exception;

class AlreadyExistsException extends \RuntimeException {}
```

> But yeah, this is used as a way to bubble information down, and to the external site, to display the message for the editors - hence why I want it to be translated.

Then you add extra properties to your exception to bubble that data. Exception messages are not meant for end users.

Exception messages tells a developer what went wrong. What you want to show your users is a different thing.

File: web/modules/custom/bnf/src/Services/BnfImporter.php [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    protected EntityTypeManagerInterface $entityTypeManager,
+    protected TranslationInterface $translation,
+    protected LoggerInterface $logger,
+  ) {}
+
+  /**
+   * Importing a node from a GraphQL source endpoint.
+   */
+  public function importNode(string $uuid, string $endpoint_url, string $node_type = 'article'): void {
+    $node_storage = $this->entityTypeManager->getStorage('node');
```
[the remark this answers is not in the document]
Comment (xendk) [reply web/modules/custom/bnf/src/Services/BnfImporter.php]: Still see a lot of snake-case.

File: phpstan.neon.dist [hunk 12]
```diff
@@ -16,6 +16,7 @@ parameters:
     # more detailed typing of.
     - '#.*\:\:(buildForm|getEditableConfigNames|submitForm|validateForm)\(\) .* no value type specified in iterable type array\.#'
     - '#_alter\(\) has parameter \$form with no value type specified in iterable type array\.#'
+    - '#_form_submit\(\) has parameter \$form with no value type specified in iterable type array\.#'
```
[the remark this answers is not in the document]
Comment (xendk) [reply phpstan.neon.dist:19]: Yes, but the commit you added it doesn't have the code that needs it. Which means that if you check that commit out and run phpstan, it'll fail. And it's confusing when reviewing by commits.

File: web/modules/custom/bnf/bnf_client/bnf_client.module (hunk 7 above)
[the remark this answers is not in the document]
Comment (xendk) [reply web/modules/custom/bnf/bnf_client/bnf_client.module:42]: Well,as it's frontend code that doesn't store any data or otherwise affect anything other, it's easy to change later. So no.

File: docker-compose.yml (hunk 10 above)
Comment (xendk) [reply docker-compose.yml:37]: > But yeah, I know that localhost should work

Ah, well no, it shouldn't. The PHP code runs in the PHP FPM container, which doesn't have a webserver. It should either be `http://nginx/graphql` (which our codes prohibits) or `https://https/graphql`, in order to actually talk with a HTTP service.

Comment (rasben) [reply docker-compose.yml]: I've found a way around it here:

https://github.com/danskernesdigitalebibliotek/dpl-cms/pull/1853/commits/65a95a4f70c7ebbdb89c086c727ce93d12462ea1

And yeah, I know that the server domain doesnt exist yet :) This was a placeholder, more than anything else.

File: config/sync/core.entity_form_display.node.article.default.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (rasben) [reply config/sync/core.entity_form_display.node.article.default.yml]: I've updated it now: https://github.com/danskernesdigitalebibliotek/dpl-cms/pull/1853/commits/3f1c7c39a69808cdd3ea05fe16d7d52b3a36ac90

Notice that I've decided to still introduce the "field" both on client and server side. This is on purpose, to avoid having to alter the code, as we both need to be able to import articles on both servers and clients.

File: docker-compose.yml (hunk 10 above)
Comment (xendk) [reply docker-compose.yml]: Well, that won't work for me, these names isn't available in the containers. I need to have a chat with @arnested...

I can't see why the `varnish` container needs this env variable?

File: config/sync/core.entity_form_display.node.article.default.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (xendk) [reply config/sync/core.entity_form_display.node.article.default.yml]: > This is on purpose, to avoid having to alter the code, as we both need to be able to import articles on both servers and clients.

What you've accomplished is that we now have two UUID fields on both sides, which people are bound to confuse. If we just use and sync node UUID we don't need to know *what* was synced, but *that* it was synced. Actually, for all intent and purposes, the syncing code doesn't even care about that, all it needs to know is that if two UUIDs match, it's the same content, and it should (potentially) update the receiver.

The only reason why we're interested in *that* it's synced is UI wise, to make lists of "stuff imported" and that can be handled with a simple `bnf_status` int. 0 = no BNF, 1 = incoming (imported from BNF client side, library submitted content server side) and 2 = outgoing (client side "send to BNF", not used server side). Code wise it's also much simpler to make a "list of imported content" by using `WHERE bnf_status = 1` than `WHERE bnf_uuid IS NOT NULL AND uuid != bnf_uuid`.

[the remark this answers is not in the document]
Comment (rasben) [reply config/sync/core.entity_form_display.node.article.default.yml]: I've introduced `bnf_state` now, as we've talked about.

File: docker-compose.yml (hunk 10 above)
Comment (rasben) [reply docker-compose.yml]: I've tweaked it a bit so it actually works for me - but, as we talked about - we need to find a solution that works for the both of us. Let's leave it here for now :)

**Review by xendk — APPROVED** [review]
:+1:
